kong/plugins/ai-prompt-decorator/handler.lua

```diff
@@ -0,0 +1,67 @@
+local _M = {}
+
+-- imports
+local kong_meta      = require "kong.meta"
+local access_handler = require("kong.plugins.ai-prompt-decorator.access")
+local re_match       = ngx.re.match
+local re_find        = ngx.re.find
+local fmt            = string.format
+local table_insert   = table.insert
+--
+
+_M.PRIORITY = 772
+_M.VERSION = kong_meta.version
+
+
+local function do_bad_request(msg)
+  kong.log.warn(msg)
+  kong.response.exit(400, { error = true, message = msg })
+end
+
+
+local function do_internal_server_error(msg)
+  kong.log.err(msg)
+  kong.response.exit(500, { error = true, message = msg })
+end
+
+
+function _M:access(conf)
+  kong.log.debug("IN: ai-prompt-decorator/access")
+  kong.service.request.enable_buffering()
+  kong.ctx.shared.prompt_decorated = true
+
+  -- if plugin ordering was altered, receive the "decorated" request
+  local request, err
+  if not kong.ctx.replacement_request then
+    request, err = kong.request.get_body("application/json")
+
+    if err then
+      do_bad_request("ai-prompt-decorator only supports application/json requests")
+    end
+  else
+    request = kong.ctx.replacement_request
+  end
```

Comment:
I think the logic can be simplified a bit. Less control flow => (usually) easier to follow along with the code:

```suggestion
  local request, err = kong.ctx.replacement_request, nil
  if not request then
    request, err = kong.request.get_body("application/json")

    if err then
      return do_bad_request("ai-prompt-decorator only supports application/json requests")
    end
  end
```

Also:

1. Let's rename `do_bad_request` => `bad_request`
2. Don't forget to `return` when calling a function that exits early by sending a response
3. I think Harry commented elsewhere about this, but I think it would be wise to leave the plugin name out of the error messages that are visible to the client. Something like `only application/json requests are supported` would be better.